### PR TITLE
Write detection logic that identifies a project as needing js/ts QP 

### LINKED ISSUE #3117: Write detection logic that identifies a project as needing js/ts QP

### PR DESCRIPTION
**Pull request by ugras-ergun-sonarsource**
fixes #3117

### CODE REVIEW

File: src/Integration/Binding/JsTsProjectTypeIndicator.cs [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+                return HasProjectJSTS(dteProject.ProjectItems);
+            }            
+        }
+
+        private bool HasFolderJSTS()
+        {
+            string root = folderWorkspaceService.FindRootDirectory();
+
+            var fileList = fileSystem.Directory.EnumerateFiles(root, "*", SearchOption.AllDirectories).Where(x => !x.Contains("\\node_modules\\"));
```
Comment (ugras-ergun-sonarsource) [inline src/Integration/Binding/JsTsProjectTypeIndicator.cs:77]: There's no overload that you can only pass the root directory and search options. So I had to pass "*" as a filter. Excluded "node_modules" to prevent having lots of extra files.

File: src/Integration/Binding/BindingProcessImpl.cs [hunk 2]
```diff
@@ -134,6 +134,11 @@ public async Task<bool> DownloadQualityProfileAsync(IProgress<FixedStepsProgress
             {
                 var serverLanguage = language.ServerLanguage;
 
+                if (serverLanguage == null)
```
Comment (ugras-ergun-sonarsource) [inline src/Integration/Binding/BindingProcessImpl.cs:137]: @rita-sonarsource this looks like your commit. Should it be there or is it some merge mistake and should be removed?

File: src/Integration/Binding/IJsTsProjectTypeIndicator.cs [hunk 3]
```diff
@@ -18,14 +18,16 @@
...
 
-namespace SonarLint.VisualStudio.Core.JsTs
+using EnvDTE;
+
+namespace SonarLint.VisualStudio.Integration.Binding
 {
     public interface IJsTsProjectTypeIndicator
     {
         /// <summary>
         /// Returns true if the currently open solution has a JS/TS file.
```
Comment (rita-sonarsource) [inline src/Integration/Binding/IJsTsProjectTypeIndicator.cs:28]: This comment needs to be updated to reflect what the new logic is doing.

File: src/Integration/Binding/BindingProcessImpl.cs (hunk 2 above)
Comment (rita-sonarsource) [reply src/Integration/Binding/BindingProcessImpl.cs:137]: Merge mistake, should be removed

File: src/Integration.UnitTests/Connection/ConnectionWorkflowTests.cs [hunk 4]
```diff
@@ -33,6 +33,7 @@
 using SonarLint.VisualStudio.Core.CFamily;
 using SonarLint.VisualStudio.Core.JsTs;
 using SonarLint.VisualStudio.Infrastructure.VS;
+using SonarLint.VisualStudio.Integration.Binding;
```
Comment (rita-sonarsource) [inline src/Integration.UnitTests/Connection/ConnectionWorkflowTests.cs:36]: Why is this file changed?

Comment (ugras-ergun-sonarsource) [reply src/Integration.UnitTests/Connection/ConnectionWorkflowTests.cs:36]: this is also a merge mistake

File: src/Integration.UnitTests/LocalServices/ErrorListInfoBarController.QualityProfileBackgroundProcessorTests.cs [hunk 5]
```diff
@@ -30,13 +30,12 @@
 using Moq;
 using SonarLint.VisualStudio.Core.Binding;
 using SonarLint.VisualStudio.Core.CFamily;
-using SonarLint.VisualStudio.Core.JsTs;
```
[the remark this answers is not in the document]
Comment (ugras-ergun-sonarsource) [reply src/Integration.UnitTests/LocalServices/ErrorListInfoBarController.QualityProfileBackgroundProcessorTests.cs:33]: it has a reference to IJsTsProjectTypeIndicator. Because we have moved it it's interface has changed.

File: src/Integration/Binding/JsTsProjectTypeIndicator.cs [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+            }
+            else
+            {
+                Debug.Assert(dteProject != null, "When it's not folder workspace we expect dteProject not to be null");
+
+                return HasProjectJsTs(dteProject.ProjectItems);
+            }            
+        }
+
+        private bool HasFolderJsTs()
```
Comment (rita-sonarsource) [inline src/Integration/Binding/JsTsProjectTypeIndicator.cs:71]: Perhaps "HasJsTsFileOnDisk" ?

File: src/Integration.UnitTests/Connection/ConnectionWorkflowTests.cs (hunk 4 above)
Comment (ugras-ergun-sonarsource) [reply src/Integration.UnitTests/Connection/ConnectionWorkflowTests.cs:36]: it has a reference to IJsTsProjectTypeIndicator. Because we have moved it, its interface has changed.

File: src/Integration/Binding/JsTsProjectTypeIndicator.cs [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+        {
+            this.sonarLanguageRecognizer = sonarLanguageRecognizer;
+            this.folderWorkspaceService = folderWorkspaceService;
+            this.fileSystem = fileSystem;
+        }
+        
+        public bool IsJsTs(Project dteProject)
+        {
+            //When opened as folder there can be a dteProject if a file is open
+            //If both are true folder search takes precedense for consistency
```
Comment (rita-sonarsource) [inline src/Integration/Binding/JsTsProjectTypeIndicator.cs:58]: "if both are true" -- both what? comment needs to be rephrased

File: src/Integration.UnitTests/Binding/JsTsProjectTypeIndicatorTests.cs [hunk 8]
```diff
@@ -0,0 +1,330 @@
...
+
+namespace SonarLint.VisualStudio.Integration.UnitTests.Binding
+{
+    [TestClass]
+    public class JsTsProjectTypeIndicatorTests
+    {
+        
+
+        [TestInitialize]
+        public void TestInitialize()
```
Comment (rita-sonarsource) [inline src/Integration.UnitTests/Binding/JsTsProjectTypeIndicatorTests.cs:25]: left over junk

File: src/Integration/Binding/JsTsProjectTypeIndicator.cs [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+                    return true;
+                }    
+            }
+            return false;
+        }
+
+        private bool IsFileJsTs(string fileName)
+        {
+            var extension = Path.GetExtension(fileName);
+            return sonarLanguageRecognizer.GetAnalysisLanguageFromExtension(extension) == Core.Analysis.AnalysisLanguage.Javascript;
```
[the remark this answers is not in the document]
Comment (ugras-ergun-sonarsource) [reply src/Integration/Binding/JsTsProjectTypeIndicator.cs:90]: Yeah, I missed that.  😔

[the remark this answers is not in the document]
Comment (rita-sonarsource) [reply src/Integration/Binding/JsTsProjectTypeIndicator.cs:90]: Please also check manually that it works.

File: src/Integration.UnitTests/Binding/JsTsProjectTypeIndicatorTests.cs [hunk 10]
```diff
@@ -0,0 +1,330 @@
...
+        {
+
+        }
+
+        [DataRow("script.js", true)]
+        [DataRow("file.cs", false)]
```
Comment (duncanp-sonar) [inline src/Integration.UnitTests/Binding/JsTsProjectTypeIndicatorTests.cs:31]: Could you add a case with a typescript file please?

File: src/Integration.UnitTests/Binding/JsTsProjectTypeIndicatorTests.cs [hunk 11]
```diff
@@ -0,0 +1,330 @@
+﻿using System;
```
Comment (duncanp-sonar) [inline src/Integration.UnitTests/Binding/JsTsProjectTypeIndicatorTests.cs:1]: Check all the new files have copyright notices.

**Review by duncanp-sonar — COMMENTED** [review]
Looks nearly there, but there need to be tests for typescript.

File: src/Infrastructure.VS/LanguageDetection/SonarLanguageRecognizer.cs [hunk 12]
```diff
@@ -112,5 +119,35 @@ private static bool IsTypeScriptDocument(IEnumerable<IContentType> contentTypes)
         {
             return contentTypes.Any(type => type.IsOfType("TypeScript"));
         }
+
+        public AnalysisLanguage? GetAnalysisLanguageFromExtension(string extension)
+        {
+            extension = NormalizeExtension(extension);
+
+            if (JavascriptSupportedExtensions.Contains(extension)) { return AnalysisLanguage.Javascript; }
```
[the remark this answers is not in the document]
Comment (duncanp-sonar) [reply src/Infrastructure.VS/LanguageDetection/SonarLanguageRecognizer.cs:127]: Probably worth adding a one-line comment to record this.

**Review by duncanp-sonar — APPROVED** [review]

File: src/Infrastructure.VS/LanguageDetection/SonarLanguageRecognizer.cs [hunk 13]
```diff
@@ -31,13 +31,20 @@ namespace SonarLint.VisualStudio.IssueVisualization.Editor.LanguageDetection
     public interface ISonarLanguageRecognizer
     {
         IEnumerable<AnalysisLanguage> Detect(string filePath, IContentType bufferContentType);
+
+        AnalysisLanguage? GetAnalysisLanguageFromExtension(string extension);
```
Comment (duncanp-sonar) [inline src/Infrastructure.VS/LanguageDetection/SonarLanguageRecognizer.cs:35]: Parameter name: we're now expecting the `fileName` rather than the `extension`, aren't we?

File: src/Integration/Binding/JsTsProjectTypeIndicator.cs [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+
+            foreach (ProjectItem item in projectItems)
+            {
+                if (IsFileJsTs(item.Name))
+                {
+                    return true;
+                }
+                if (item.ProjectItems?.Count > 0)
+                {
+                    if (HasJsTsFileInProject(item.ProjectItems))
```
Comment (duncanp-sonar) [inline src/Integration/Binding/JsTsProjectTypeIndicator.cs:105]: SQ raised a code smell for this nested `if` - could be a single `if`.

File: src/Integration/Binding/JsTsProjectTypeIndicator.cs [hunk 15]
```diff
@@ -0,0 +1,115 @@
...
+            }            
+        }
+
+        private bool HasJsTsFileOnDisk()
+        {
+            string root = folderWorkspaceService.FindRootDirectory();
+
+            var fileList = fileSystem.Directory.EnumerateFiles(root, "*", SearchOption.AllDirectories).Where(x => !x.Contains("\\node_modules\\"));
+
+            foreach (var file in fileList)
```
Comment (duncanp-sonar) [inline src/Integration/Binding/JsTsProjectTypeIndicator.cs:78]: SQ raised a code smell suggesting using LINQ to simplify this expression.
Something like `return (fileList.Any(IsFileJsTs))` maybe?

**Comment by duncanp-sonar** [comment]
@ugras-ergun-sonarsource don't forget to check the [PR issues](https://github.com/SonarSource/sonarlint-visualstudio/pull/3120/checks?check_run_id=7536170888) raised by SonarCloud.

File: src/Infrastructure.VS/LanguageDetection/SonarLanguageRecognizer.cs [hunk 16]
```diff
@@ -112,5 +119,37 @@ private static bool IsTypeScriptDocument(IEnumerable<IContentType> contentTypes)
         {
             return contentTypes.Any(type => type.IsOfType("TypeScript"));
         }
+
+        public AnalysisLanguage? GetAnalysisLanguageFromExtension(string fileName)
+        {
+            var extension = GetNormalizedExtention(fileName);
+
+            if (JavascriptSupportedExtensions.Contains(extension)) { return AnalysisLanguage.Javascript; }
+            if (IsContentType(TypeScriptTypeName, extension)) { return AnalysisLanguage.TypeScript; }
```
Comment (rita-sonarsource) [inline src/Infrastructure.VS/LanguageDetection/SonarLanguageRecognizer.cs:128]: What happens if a user opens a folder with a .ts file in it? I don't think it will have a content type.

File: src/Infrastructure.VS/LanguageDetection/SonarLanguageRecognizer.cs [hunk 17]
```diff
@@ -112,5 +119,37 @@ private static bool IsTypeScriptDocument(IEnumerable<IContentType> contentTypes)
...
+                //remove the leading . on extension
+                extension = extension.Substring(1);
+            }
+            return extension.ToLowerInvariant();
+        }
+
+        private bool IsContentType(string typeName, string extension)
+        {
+            var contentType = contentTypeRegistryService.GetContentType(typeName);
+            var extensions = fileExtensionRegistryService.GetExtensionsForContentType(contentType);
```
Comment (rita-sonarsource) [inline src/Infrastructure.VS/LanguageDetection/SonarLanguageRecognizer.cs:150]: I saw in the API that `fileExtensionRegistryService` has a method GetContentTypeForExtension.
Wouldn't that simplify? https://docs.microsoft.com/en-us/dotnet/api/microsoft.visualstudio.utilities.ifileextensionregistryservice.getcontenttypeforextension?view=visualstudiosdk-2022#microsoft-visualstudio-utilities-ifileextensionregistryservice-getcontenttypeforextension(system-string)

File: src/Integration/Binding/IJsTsProjectTypeIndicator.cs [hunk 18]
```diff
@@ -18,14 +18,16 @@
...
+using EnvDTE;
+
+namespace SonarLint.VisualStudio.Integration.Binding
 {
     public interface IJsTsProjectTypeIndicator
     {
         /// <summary>
-        /// Returns true if the currently open solution has a JS/TS file.
-        /// Returns false if the solution has no JS/TS files, or if there is no open solution.
+        /// Returns true if the currently open solution/folder has a JS/TS file.
```
Comment (rita-sonarsource) [inline src/Integration/Binding/IJsTsProjectTypeIndicator.cs:28]: The word "solution" is misleading here. We're not checking the solution but the given project. There is also no mention of what happens if the project is null and we are in open folder/not in open as folder.

File: src/Infrastructure.VS/LanguageDetection/SonarLanguageRecognizer.cs (hunk 16 above)
Comment (ugras-ergun-sonarsource) [reply src/Infrastructure.VS/LanguageDetection/SonarLanguageRecognizer.cs:128]: We do not check the file's content type but if the file extension is associated with the given type.

**Review by duncanp-sonar — COMMENTED** [review]
See inline comments.

File: src/Integration/Binding/JsTsProjectTypeIndicator.cs (hunk 14 above)
Comment (ugras-ergun-sonarsource) [reply src/Integration/Binding/JsTsProjectTypeIndicator.cs:105]: I have seen it but I think it would reduce the readability. We go to the inner loop only if the item has child items and if we merge it, this won't be obvious.

File: src/Infrastructure.VS/LanguageDetection/SonarLanguageRecognizer.cs [hunk 19]
```diff
@@ -112,5 +119,34 @@ private static bool IsTypeScriptDocument(IEnumerable<IContentType> contentTypes)
         {
             return contentTypes.Any(type => type.IsOfType("TypeScript"));
         }
+
+        public AnalysisLanguage? GetAnalysisLanguageFromExtension(string fileName)
+        {
+            var extension = GetNormalizedExtention(fileName);
+
+            var ctype = fileExtensionRegistryService.GetContentTypeForExtension(extension);
```
Comment (duncanp-sonar) [inline src/Infrastructure.VS/LanguageDetection/SonarLanguageRecognizer.cs:127]: @ugras-ergun-sonarsource  SonarCloud is complaining that this variable is not used.

Comment (ugras-ergun-sonarsource) [reply src/Infrastructure.VS/LanguageDetection/SonarLanguageRecognizer.cs]: I've already removed it and pushed.

**Review by rita-sonarsource — APPROVED** [review]
@ugras-ergun-sonarsource LGTM but I haven't seen your reply to this comment: https://github.com/SonarSource/sonarlint-visualstudio/pull/3120#discussion_r930923133

**Comment by ugras-ergun-sonarsource** [comment]
> @ugras-ergun-sonarsource LGTM but I haven't seen your reply to this comment: [#3120 (comment)](https://github.com/SonarSource/sonarlint-visualstudio/pull/3120#discussion_r930923133)

I have used the GetContentTypeForExtensionthat you have recommended
